packages/replay-next/src/suspense/SourceHitCountsCache.ts

```diff
@@ -0,0 +1,99 @@
+import { toPointRange } from "shared/utils/time";
+
+import { breakpointPositionsCache } from "./BreakpointPositionsCache";
+
+type MinMaxHitCountTuple = [minHitCount: number, maxHitCount: number];
+
+// Source id and focus range to a tuple of min and max hit counts;
+// this value is updated as we fetch new hit counts for the source (and focus range)
+const minMaxHitCountsMap: Map<string, MinMaxHitCountTuple> = new Map();
+
+function getKey(
+  replayClient: ReplayClientInterface,
+  sourceId: SourceId,
+  focusRange: PointRange | null
+) {
+  return focusRange ? `${sourceId}:${focusRange.begin}-${focusRange.end}` : sourceId;
+}
+
+export const sourceHitCountsCache = createIntervalCache<
+  number,
+  [replayClient: ReplayClientInterface, sourceId: SourceId, focusRange: PointRange | null],
+  [number, LineHitCounts]
+>({
+  debugLabel: "SourceHitCountsCache",
+  getPointForValue: ([line]) => line,
+  comparePoints: (a, b) => a - b,
+  getKey,
+  load: async (begin, end, client, sourceId, focusRange) => {
+    try {
+      const [locations] = await breakpointPositionsCache.readAsync(client, sourceId);
+      const hitCounts = await client.getSourceHitCounts(
+        sourceId,
+        {
+          start: { line: begin, column: 0 },
+          end: { line: end, column: Number.MAX_SAFE_INTEGER },
+        },
+        locations,
+        focusRange
+      );
+
+      // Refine cached min-max hit count as we load more information about a source.
+      // Because hit counts may be filtered by a range of lines, this value may change as new information is loaded in.
+      const key = getKey(client, sourceId, focusRange);
+      let [minHitCount, maxHitCount] = minMaxHitCountsMap.get(key) || [Number.MAX_SAFE_INTEGER, 0];
+      hitCounts.forEach(hitCount => {
+        const { count } = hitCount;
+        minHitCount = Math.min(minHitCount, count);
+        maxHitCount = Math.max(maxHitCount, count);
+      });
+      minMaxHitCountsMap.set(key, [minHitCount, maxHitCount]);
+
+      return [...hitCounts.entries()];
+    } catch (error) {
+      if (
+        !isCommandError(error, ProtocolError.TooManyLocationsToPerformAnalysis) &&
+        !isCommandError(error, ProtocolError.LinkerDoesNotSupportAction)
+      ) {
+        throw error;
+      }
+      return [];
+    }
+  },
+});
+
+export function useGetSourceHitCounts(
+  client: ReplayClientInterface,
+  sourceId: SourceId,
+  locationRange: SourceLocationRange,
+  focusRange: TimeStampedPointRange | null
+) {
+  const begin = locationRange.start.line;
+  const end = locationRange.end.line;
+  const pointRange = focusRange ? toPointRange(focusRange) : null;
+  sourceHitCountsCache.readAsync(begin, end, client, sourceId, pointRange);
+  const status = useIntervalCacheStatus(
+    sourceHitCountsCache,
+    begin,
+    end,
+    client,
+    sourceId,
+    pointRange
+  );
+  if (status === "resolved") {
+    return sourceHitCountsCache.read(begin, end, client, sourceId, pointRange);
+  }
+}
```

Comment:
If I'm understanding it correctly– the only reason it exists is to let hit counts be loaded _without_ using Suspense? (else we'd be using `sourceHitCountsCache.read` instead of `sourceHitCountsCache.readAsync`)

I don't think we need it. Should be able to just use `sourceHitCountsCache` directly.

Reply:
> If I'm understanding it correctly– the only reason it exists is to let hit counts be loaded without using Suspense? 

Correct.

> I don't think we need it. Should be able to just use sourceHitCountsCache directly.

Yes, that would be possible, but this is a performance optimization we consciously added in #8695.

Reply:
I see. So the goal here is not so much to bypass Suspense– but to complete the initial render _without suspending_ if data hasn't let been loaded. The `useImperativeCacheValue` hook (in "suspense") does this, although that's not its main purpose– and it doesn't work with interval caches.

Maybe I should just add a `useIntervalCacheValue` hook to the "suspense" package too.

Reply:
> Maybe I should just add a useIntervalCacheValue hook to the "suspense" package too.

Yes, I think that would make sense.

Reply:
Oh it.

Reply:
PR #8968 updates `suspense@0.0.28` which adds a new hook, `useImperativeIntervalCacheValues`:
https://suspense.vercel.app/useImperativeIntervalCacheValues

Should be usable here 😄

Reply:
I'll rebase when that PR is fixed and merged.

Reply:
Should be ready now!